cf-check/validate.c

```diff
@@ -0,0 +1,515 @@
+#include <platform.h>
+#include <validate.h>
+#include <logging.h>
+
+#if defined(__MINGW32__) || !defined(LMDB)
+
+int CFCheck_Validate(const char *path);
+{
+    Log(LOG_LEVEL_ERR,
+        "cf-check diagnose --validate not available on this platform/build");
+    return false;
+}
+
+#else
+
+// #include <stdio.h>
+// #include <lmdump.h>
+#include <lmdb.h>       // mdb_open(), mdb_close(), mdb_strerror()
+// #include <sys/wait.h>
+// #include <signal.h>
+// #include <utilities.h>
+// #include <sequence.h>
+#include <alloc.h>      // xcalloc(), xmemdup(), xstrdup()
+#include <string_lib.h> // StringEndsWith()
+// #include <unistd.h>
+#include <map.h>        // StringMap
+#include <set.h>        // StringSet
+
+static int lmdb_report_error(int rc)
+{
+    printf("err(%d): %s\n", rc, mdb_strerror(rc));
+    return rc;
+}
+
+typedef enum ValidatorMode
+{
+    CF_CHECK_VALIDATE_UNKNOWN,
+    CF_CHECK_VALIDATE_LASTSEEN,
+} ValidatorMode;
+
+typedef struct LastSeenState
+{
+    StringMap *hostkey_to_address;
+    StringMap *address_to_hostkey;
+    StringSet *quality_outgoing_hostkeys;
+    StringSet *quality_incoming_hostkeys;
+} LastSeenState;
+
+typedef struct ValidatorState
+{
+    const char *path;
+    ValidatorMode mode;
+    size_t errors;
+    StringSet *keys;
+    Seq *values;
+    union
+    {
+        LastSeenState lastseen;
+    };
+} ValidatorState;
+
+static Slice *NewValue(void *data, size_t size)
+{
+    assert(data != NULL);
+    assert(size > 0);
+
+    Slice *r = xcalloc(1, sizeof(Slice));
+    r->size = size;
+    r->data = xmemdup(data, size);
+
+    return r;
+}
+
+static void DestroyValue(Slice *value)
```

Comment:
I'd use the word `Slice` in the names of these two functions.

Reply:
But the idea is that this is really not the general New/Destroy methods for slices, they are specific to these LMDB values.

Reply:
Renamed to `DestroyLMDBSlice()` and `NewLMDBSlice()`.